yeoman-custom/cli/man/yeoman.1
.\" Generated with Ronnjs 0.3.8
.\" http://github.com/kapouer/ronnjs/
.
.TH "YEOMAN" "1" "July 2012" "" ""
.
.SH "NAME"
\fByeoman\fR \-\- Yeoman CLI tool
.
.SH "What am I?"
Yeoman is a robust and opinionated client\-side stack, comprised of tools and
frameworks that can help developers quickly build beautiful web applications\.
We take care of providing everything needed to get started without any of the
normal headaches associated with a manual setup\.
.
.P
Yeoman is fast, performant and is optimized to work best in modern browsers\.
.
.P
For more information about the project, see \fIhttp://yeoman\.io\fR\|\.
.
.SH "Installing"
.
.IP "\(bu" 4
Clone this repo and \fBcd\fR into it
.
.IP "\(bu" 4
Run the following command at the terminal:
.
.IP "" 0
.
.P
\fBshell
\|\./setup/install\.sh\fR
.
.IP "\(bu" 4
Navigate to a new directory and run \fByeoman init\fR to make sure everything is
working as expected\.
.
.IP "" 0
.
.SS "Trouble\-shooting"
If for any reason you experience exceptions after the yeoman installation
process above, you may find the following steps resolve these issues:
.
.P
\fB
$ cd yeoman/cli
$ sudo npm install \-g
# when complete then run\.\.
$ sudo npm link\fR
.
.SH "Running"
Here\'s a small shell script that you can save as \fBserver\.sh\fR which opens and
servers the current directory on the port specified:
.
.P
\fB\fR`shell
port=$1
if [ $#  \-ne  1 ]
then
  port=8000
fi
.
.P
if [ $(uname \-s) == "Darwin" ]
then
  open=open
else
  open=xdg\-open
fi
.
.P
$open http://localhost:$port && python \-m SimpleHTTPServer $port; \fB\fR`
.
.P
For example, run this guy as:
.
.P
\fBshell
\|\./server\.sh 8000\fR
.
.SH "Documentation"
The current documentation for Yeoman can be found here \fIhttp://yeoman\.github\.com/docs\fR\|\. If you are a new contributor and require
access to this repository, feel free to ask\.
.
.SH "Browser Support"
Yeoman supports:
.
.IP "\(bu" 4
Modern browsers (latest versions of Chrome, Safari, Firefox, Opera and IE10)
.
.IP "\(bu" 4
Chrome on Android
.
.IP "\(bu" 4
Mobile Safari
.
.IP "" 0
.
.SH "Contribute"
.
.SS "Repos"
.
.IP "\(bu" 4
Yeoman (CLI, Insights) \fIhttp://github\.com/yeoman/yeoman\fR
.
.IP "\(bu" 4
Yeoman I/O Holding Page \fIhttp://github\.com/yeoman/yeoman\.io\fR
.
.IP "\(bu" 4
Yeoman I/O Site \fIhttp://github\.com/yeoman/yeoman\.io\fR(site branch)
.
.IP "\(bu" 4
Yeoman Docs \fIhttp://github\.com/yeoman/docs\fR
.
.IP "" 0
.
.SS "Style Guide"
This project follows the jQuery Style
Guide \fIhttp://docs\.jquery\.com/JQuery_Core_Style_Guidelines\fR with an exception
of two space indentation and multiple var statements\. Please ensure any pull
requests follow this closely\. If you notice existing code which doesn\'t follow
these practices, feel free to shout and we will address this\.
.
.SH "About"
Yeoman is an open\-source project by Google which builds on top of a number of
open\-source solutions\. These include Grunt, Twitter Bootstrap and Compass\.
Version 1 of the project features the combined efforts of:
.
.IP "\(bu" 4
Paul Irish
.
.IP "\(bu" 4
Addy Osmani
.
.IP "\(bu" 4
Mickael Daniel
.
.IP "\(bu" 4
Sindre Sorhus
.
.IP "\(bu" 4
Eric Bidelman \fIhttp://ericbidelman\.com\fR
.
.IP "" 0
.
.P
and other developers\.
.
.P
We will be aiming to officially release the project in late July, 2012\.


changelog.md
### 0.9.5 - 2012-11-05

- [#627](https://github.com/yeoman/yeoman/issues/627)
Fix: Usemin:html adds image file before a 'a' anchor

- [#636](https://github.com/yeoman/yeoman/issues/636)
Python3 support for Insight

- [#603](https://github.com/yeoman/yeoman/issues/603)
Add ability to specify a base path in the usemin block comment

- [#660](https://github.com/yeoman/yeoman/issues/660)
Determine linefeed from the content not the platform

- Upgrade to latest Bower


### 0.9.4 (12 October, 2012)

* Updated to Bower 0.2.0 (Issue [#551](https://github.com/yeoman/yeoman/issues/551))
* Less verbose `yeoman init` (Issue [#605](https://github.com/yeoman/yeoman/issues/605))
* Ability to disable Insight (Issue [#305](https://github.com/yeoman/yeoman/issues/385))
* Corrections to generator docs (Issue [#608](https://github.com/yeoman/yeoman/issues/608))
* Fixes to usemin:html skipping anchor tags with images as href (Issue [#615](https://github.com/yeoman/yeoman/issues/615))
* Fixes to main.css not being replaced with compiled version unless tag structure is exactly as yeoman expects it (Issue [#502](https://github.com/yeoman/yeoman/issues/502))
* Stops yeoman from looking for win32 binaries in ../vendor/ (Pull [#519](https://github.com/yeoman/yeoman/pull/519))
* Support for installing Yeoman behind a corporate proxy server (Pull [#587](https://github.com/yeoman/yeoman/pull/587))
* Remove a trailing comma in the options of the HTML task and add missing semicolons in the server task
(Issue [#589](https://github.com/yeoman/yeoman/pull/589))
* Fix to enable blank lines in usemin blocks (Issue [#560](https://github.com/yeoman/yeoman/issues/560))
* Updates to Insight documentation
* Fixes usemin:css inserting invalid <link> element (Issue [#586](https://github.com/yeoman/yeoman/issues/586))
* Fix for usemin replace task incorrectly matcheing filename when files in subfolders have the same name (Issue [#565](https://github.com/yeoman/yeoman/issues/565))
* New Windows [installation](https://github.com/yeoman/yeoman/wiki/Manual-Install) instructions 

### 0.9.3 (3 October, 2012)

* Temporary fix to allow Bower dependencies to be correctly copied to app/components. As a result of this patch RequireJS wiring of Bower deps is currently disabled

### 0.9.2 

TBU

### 0.9.1 (15 September, 2012)

* Removal of stdout checks, test against fs (Pull [#473](https://github.com/yeoman/yeoman/pull/473)).

* Fix for  initializer example (Pull [#447](https://github.com/yeoman/yeoman/pull/477 )).

* Support for an expanded compass --require arg (Pull [#483](https://github.com/yeoman/yeoman/pull/483)).   

* Yeoman setup gets a colored art (Pull [#493](https://github.com/yeoman/yeoman/pull/493))!

* Added ability to configure app index file for rjs task (Pull [#505](https://github.com/yeoman/yeoman/pull/505)).

* yeoman server:test and access to the app js files (Issue [#443](https://github.com/yeoman/yeoman/issues/443)).

* yeoman install backbone deletes existing files in app/scripts/vendor issue was fixed (Issue [#460](https://github.com/yeoman/yeoman/issues/460)).

* Replaced coffescript task with grunt-coffee (Pull[#522](https://github.com/yeoman/yeoman/pull/522))


cli/tasks/yeoman.js
var fs    = require('fs');
var join  = require('path').join;
var util  = require('util');
var which = require('which');

// ant build script has a nice notion of environment, this defaults to
// production. And we only support env=prod for now.
//
// not implemented tasks (add noop waithing for their impl): manifest

module.exports = function(grunt) {

  // External grunt plugin:
  //
  // - jasmine task: https://github.com/creynders/grunt-jasmine-task
  //
  // note: We need to use loadTasks instead of loadNpmTasks, otherwise will try
  // to load relative to gruntfile node_modules, this would require user to
  // install manually. So we load tasks specifically from our node_modules,
  // with abs path.
  //
  grunt.loadTasks(join(__dirname, '../node_modules/grunt-jasmine-task/tasks'));
  grunt.loadTasks(join(__dirname, '../node_modules/grunt-mocha/tasks'));
  grunt.loadTasks(join(__dirname, '../node_modules/grunt-contrib-coffee/tasks'));

  // build targets: these are equivalent to grunt alias except that we defined
  // a single task and use arguments to trigger the appropriate target
  //
  // - build    - no html compression, no usemin-handler task
  // - usemin   - (default) same as build but parsing config from markup
  // - text     - same as usemin but without image (png/jpg) optimizing
  // - buildkit - minor html optimizations, all html whitespace/comments
  //              maintained (todo: inline script/style minified)
  // - basics   - same as buildkit plus minor html optimizations
  // - minify   - same as build plus full html minification
  //
  // - test     - same as default build plus but conditionally runs compass /
  //              manifest task depending on whether or not compass / phantomjs binaries are
  //              available within the path.  During the checking process, we output warning
  //              infos about missing deps. It might make sense to make it the default (img
  //              task internally does this check)
  var targets = {
    default   : '               rjs concat css min img rev usemin manifest',
    usemin    : 'usemin-handler rjs concat css min img rev usemin manifest',
    text      : 'usemin-handler rjs concat css min     rev usemin manifest',
    buildkit  : 'usemin-handler rjs concat css min img rev usemin manifest html:buildkit',
    basics    : 'usemin-handler rjs concat css min img rev usemin manifest html:basics',
    minify    : 'usemin-handler rjs concat css min img rev usemin manifest html:compress',
    test      : 'usemin-handler rjs concat css min img rev usemin manifest'
  };

  var targetList = grunt.log.wordlist(Object.keys(targets));
  grunt.registerTask('build', 'Run a predefined target - build:<target> \n' + targetList, function(target) {
    var valid = Object.keys(targets);
    target = target || 'usemin';

    if ( valid.indexOf( target ) === -1 ) {
      grunt.log
        .error('Not a valid target')
        .error(grunt.helper('invalid targets', targets));
      return false;
    }

    var tasks = [
      'intro',
      grunt.config('build.tasks._beforeAll') || 'clean coffee compass mkdirs',
      grunt.config('build.tasks.' + target) || targets[target],
      grunt.config('build.tasks._afterAll'),
      'copy time'
    ].join(' ');

    // Conditionally remove compass / manifest task if either compass or
    // phantomjs binary is missing. Done only for `test` target (specifically
    // used for our `npm test`). For each, output warning infos.
    if( target === 'test' ) {
      tasks = grunt.helper( 'build:skip', tasks, 'compass' );
      tasks = grunt.helper( 'build:skip', tasks, 'phantomjs', 'manifest' );
    }

    grunt.log.subhead('Running ' + target + ' target')
      .writeln('  - ' + grunt.log.wordlist(tasks.split(' '), { separator: ' ' }));


    grunt.task.run(tasks);
  });

  grunt.registerHelper('invalid targets', function(valid, code) {
    var msg = Object.keys(valid).map(function(key) {
      if ( /pre|post/.test( key ) ) {
        return '';
      }
      return grunt.helper('pad', key, 10) + '# '+ valid[key];
    }).join(grunt.util.linefeed);

    var err = new Error(grunt.util.linefeed + msg);
    err.code = code || 3;
    return err;
  });

  // Helper to conditionally remove a given task from the provided list of
  // `tasks` (as a String). A `cmd` and `task` Strings should be provided,
  // if `task` is omitted, it then takes the same value of the `cmd` one.
  //
  // When a binary is known as non available, and that the task is skipped,
  // this helper also output warning info to the user.
  //
  // - tasks   - String list of tasks to be run
  // - cmd     - The String name of the binary to check against the system.
  // - task    - (optional) A string name to define the specific task to remove
  //             from the returned tasks list. If undefined, is the same than
  //             `binary`.
  //
  // Returns optionally modified `tasks` String.
  grunt.registerHelper('build:skip', function(tasks, cmd, task) {
      try {
        which.sync(cmd);
      } catch(err) {
        grunt.helper('not installed', cmd);
        tasks = tasks.replace(task || cmd, '');
      }

      return tasks;
  });

  grunt.registerHelper('pad', function pad(str, max) {
    return str.length > max ? str :
        str + new Array(max - str.length + 1).join(' ');
  });

  var now = +new Date();
  grunt.registerTask('time', 'Print sucess status with elapsed time', function() {
    grunt.log.ok('Build success. Done in ' + ((+new Date() - now) / 1000) + 's');
  });

  // Output some size info about a file, from a stat object.
  grunt.registerHelper('min_max_stat', function(min, max) {
    min = typeof min === 'string' ? fs.statSync(min) : min;
    max = typeof max === 'string' ? fs.statSync(max) : max;
    grunt.log.writeln('Uncompressed size: ' + String(max.size).green + ' bytes.');
    grunt.log.writeln('Compressed size: ' + String(min.size).green + ' bytes minified.');
  });

  // Output some info on given object, using util.inspect, using colorized output.
  grunt.registerHelper('inspect', function(o) {
    return util.inspect(o, false, 4, true);
  });

};
